File: solutions/java/go-counting/1/src/main/java/GoCounting.java
import java.awt.Point;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class GoCounting {
    private final String[] rows;
    private final int width;
    private final int height;

    GoCounting(String board) {
        if (board == null || board.isEmpty()) {
            this.rows = new String[0];
            this.width = 0;
            this.height = 0;
        } else {
            this.rows = board.split("\n");
            this.height = this.rows.length;
            this.width = this.height > 0 ? this.rows[0].length() : 0;
        }
    }

    private void checkBounds(int x, int y) {
        if (x < 0 || x >= width || y < 0 || y >= height) {
            throw new IllegalArgumentException("Invalid coordinate");
        }
    }

    Player getTerritoryOwner(int x, int y) {
        checkBounds(x, y);
        // A stone is not part of any territory
        if (rows[y].charAt(x) != ' ') {
            return Player.NONE;
        }
        
        Set<Point> territory = getTerritory(x, y);
        boolean touchesBlack = false;
        boolean touchesWhite = false;
        
        // Check all neighbors of all points in the territory
        for (Point p : territory) {
            int[] dx = {0, 0, 1, -1};
            int[] dy = {1, -1, 0, 0};
            for (int i = 0; i < 4; i++) {
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];
                if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
                    char neighbor = rows[ny].charAt(nx);
                    if (neighbor == 'B') touchesBlack = true;
                    if (neighbor == 'W') touchesWhite = true;
                }
            }
        }
        
        // Determine owner based on bordering stones
        if (touchesBlack && !touchesWhite) return Player.BLACK;
        if (touchesWhite && !touchesBlack) return Player.WHITE;
        return Player.NONE;
    }

    Set<Point> getTerritory(int x, int y) {
        checkBounds(x, y);
        // A stone is not part of any territory
        if (rows[y].charAt(x) != ' ') {
            return Collections.emptySet();
        }
        
        Set<Point> territory = new HashSet<>();
        Queue<Point> queue = new LinkedList<>();
        queue.add(new Point(x, y));
        territory.add(new Point(x, y));
        
        // BFS to find all connected empty intersections
        while (!queue.isEmpty()) {
            Point p = queue.poll();
            int[] dx = {0, 0, 1, -1};
            int[] dy = {1, -1, 0, 0};
            
            for (int i = 0; i < 4; i++) {
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];
                
                if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
                    char neighbor = rows[ny].charAt(nx);
                    if (neighbor == ' ' && !territory.contains(new Point(nx, ny))) {
                        territory.add(new Point(nx, ny));
                        queue.add(new Point(nx, ny));
                    }
                }
            }
        }
        return territory;
    }

    Map<Player, Set<Point>> getTerritories() {
        Map<Player, Set<Point>> territories = new HashMap<>();
        territories.put(Player.BLACK, new HashSet<>());
        territories.put(Player.WHITE, new HashSet<>());
        territories.put(Player.NONE, new HashSet<>());

        boolean[][] visited = new boolean[height][width];

        // Scan the entire board to find all disconnected territories
        for (int r = 0; r < height; r++) {
            for (int c = 0; c < width; c++) {
                if (!visited[r][c] && rows[r].charAt(c) == ' ') {
                    Set<Point> currentTerritory = new HashSet<>();
                    Queue<Point> queue = new LinkedList<>();
                    queue.add(new Point(c, r));
                    visited[r][c] = true;
                    
                    boolean touchesBlack = false;
                    boolean touchesWhite = false;
                    
                    // BFS for the current territory
                    while (!queue.isEmpty()) {
                        Point p = queue.poll();
                        currentTerritory.add(p);
                        
                        int[] dx = {0, 0, 1, -1};
                        int[] dy = {1, -1, 0, 0};
                        
                        for (int i = 0; i < 4; i++) {
                            int nx = p.x + dx[i];
                            int ny = p.y + dy[i];
                            
                            if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
                                char neighbor = rows[ny].charAt(nx);
                                if (neighbor == 'B') touchesBlack = true;
                                else if (neighbor == 'W') touchesWhite = true;
                                else if (neighbor == ' ' && !visited[ny][nx]) {
                                    visited[ny][nx] = true;
                                    queue.add(new Point(nx, ny));
                                }
                            }
                        }
                    }
                    
                    // Assign the territory to the correct player
                    Player owner;
                    if (touchesBlack && !touchesWhite) owner = Player.BLACK;
                    else if (touchesWhite && !touchesBlack) owner = Player.WHITE;
                    else owner = Player.NONE;
                    
                    territories.get(owner).addAll(currentTerritory);
                }
            }
        }
        return territories;
    }
}

File: solutions/javascript/go-counting/1/go-counting.js
export class GoCounting {
  constructor(board) {
    this.board = board;
    this.height = board.length;
    this.width = this.height > 0 ? board[0].length : 0;
  }

  getTerritory(x, y) {
    // 1. Validate coordinates
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return { error: 'Invalid coordinate' };
    }

    // 2. If the starting point is a stone, it has no empty territory
    if (this.board[y][x] !== ' ') {
      return { owner: 'NONE', territory: [] };
    }

    const visited = new Set();
    const territory = [];
    const queue = [[x, y]];
    const boundingStones = new Set();

    visited.add(`${x},${y}`);

    // 3. Breadth-First Search to map the empty territory
    while (queue.length > 0) {
      const [cx, cy] = queue.shift();
      territory.push([cx, cy]);

      // Orthogonal neighbors (left, right, up, down)
      const neighbors = [
        [cx - 1, cy],
        [cx + 1, cy],
        [cx, cy - 1],
        [cx, cy + 1],
      ];

      for (const [nx, ny] of neighbors) {
        // Ensure neighbor is within board boundaries
        if (nx >= 0 && nx < this.width && ny >= 0 && ny < this.height) {
          const char = this.board[ny][nx];
          
          if (char === ' ') {
            const key = `${nx},${ny}`;
            if (!visited.has(key)) {
              visited.add(key);
              queue.push([nx, ny]);
            }
          } else {
            // It's a stone; record its color to determine ownership later
            boundingStones.add(char);
          }
        }
      }
    }

    // 4. Determine ownership based on surrounding stones
    let owner = 'NONE';
    if (boundingStones.has('B') && !boundingStones.has('W')) {
      owner = 'BLACK';
    } else if (boundingStones.has('W') && !boundingStones.has('B')) {
      owner = 'WHITE';
    }

    // Sort coordinates first by X (column), then by Y (row) for predictable testing
    territory.sort((a, b) => (a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]));

    return { owner, territory };
  }

  getTerritories() {
    const result = {
      territoryBlack: [],
      territoryWhite: [],
      territoryNone: [],
    };

    const globalVisited = new Set();

    // Scan the board sequentially to find unmapped territories
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const char = this.board[y][x];
        const key = `${x},${y}`;

        if (char === ' ' && !globalVisited.has(key)) {
          const { owner, territory } = this.getTerritory(x, y);
          
          // Mark all cells in this newly discovered territory as visited globally
          for (const [tx, ty] of territory) {
            globalVisited.add(`${tx},${ty}`);
          }

          // Push the found territory into the correct property
          if (owner === 'BLACK') {
            result.territoryBlack.push(...territory);
          } else if (owner === 'WHITE') {
            result.territoryWhite.push(...territory);
          } else {
            result.territoryNone.push(...territory);
          }
        }
      }
    }

    return result;
  }
}
